### PR TITLE
Implement loading state / video buffering

### PR DESCRIPTION
**Pull request by christophblessing**
## What does this change?

This PR implements a loading state indicator.

https://user-images.githubusercontent.com/22977799/223768467-78bc7222-476b-40b5-82b1-b26fe136d944.mov

- [x] Show spinner (similar to https://fontawesome.com/icons/spinner-third?s=regular&f=classic)
- [x] Add delay to the spinner (not showing in the first 250ms or so)
- [x] Show a loading indicator for all/both video streams
- [x] Colored white with a dark ~~(drop) shadow~~ background to be visible on the video when buffering

See XI-5568

## Decisions / Choices I made

The loading state is initially true. After the player has initialized, it's set to false.

Further implementation is based on events that the video providers expose.

These are:

1) [`bufferstart`](https://developer.vimeo.com/player/sdk/reference#bufferstart) and [`buffered`](https://developer.vimeo.com/player/sdk/reference#bufferend) and  for Vimeo

~~You see no changes for the Vimeo provider as we already implemented but never used these events.~~

Edit: In the meantime, Vimeo decided to implement their own buffering indicators.
They are not hidden by the embed option `controls`, because it's not an interactive element.
Since the video is embedded as an `iframe`, we cannot change styles or logic have no other choice than to use it.
Therefore, the buffering event handling was removed for the Vimeo provider.

2) [`playing`](https://github.com/kaltura/playkit-js/blob/master/src/event/event-type.js#L57) and [`waiting`](https://github.com/kaltura/playkit-js/blob/master/src/event/event-type.js#L93) for Kaltura

They are exposed as the internal events `buffering` and `buffered` form the `XmVideo` interface.
I did not change the names as they were already there.

The player `@Listen`s to these events and changes the state `loading` accordingly.
This leads to the rendering of a new component called `xm-loading`,
which encapsulates the styles.

## Further changes

- There is now a system for `z-index` that is used project-wide
- Added video player example with Kaltura video provider

### CODE REVIEW

File: src/components/state/loading.scss [hunk 1]
```diff
@@ -0,0 +1,38 @@
+.loading {
+  animation: 0s linear 750ms forwards delayedShow;
+  visibility: hidden;
+
+  position: absolute;
+  inset: 0;
+  margin-bottom: 47px;
```
Comment (christophblessing) [inline src/components/state/loading.scss:7]: We need to change this margin according how high the control toolbar is. In full screen mode, there is a gap between overlay and toolbar.

Comment (JuliaCasamitjana) [reply src/components/state/loading.scss:7]: what about moving the component inside of` xm-screen`?

**Comment by JuliaCasamitjana** [comment]
Should we implement in this MR the suggestion to pause the streams while buffering (and restoring the state afterward)?

File: src/components/state/loading.scss (hunk 1 above)
Comment (christophblessing) [reply src/components/state/loading.scss:7]: This sound reasonable, but we only render the screen component [if there are two videos side by side](https://github.com/openHPI/xikolo-video-player/blob/main/src/components/player/player.tsx#L757). The naming `xm-screen` is somewhat misleading.

Comment (JuliaCasamitjana) [reply src/components/state/loading.scss:7]: right. Let's adjust this margin then or add a fix height for the controls.

File: src/components/state/loading.scss [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  width: 100%;
+  background-color: rgba(0, 0, 0, 0.5);
+
+  z-index: 10;
```
[the remark this answers is not in the document]
Comment (christophblessing) [reply src/components/state/loading.scss]: Resolved in: deb661fc9b2a5e7dde17986cd809ae74b237e251

File: src/utils/status.ts [hunk 3]
```diff
@@ -37,6 +37,7 @@ export interface Status {
   progress: Progress;
   subtitle: Subtitle;
   settings: Settings;
+  loading: Boolean;
```
[the remark this answers is not in the document]
Comment (JuliaCasamitjana) [reply src/utils/status.ts]: done ✅

File: src/components/state/loading.scss (hunk 1 above)
Comment (JuliaCasamitjana) [reply src/components/state/loading.scss]: I've adjusted the margin [here](https://github.com/openHPI/xikolo-video-player/pull/41/commits/87fe619b4776cfd1348cdbe1f8857f18b3ded2c8). For that, the status needs to be passed to the component.

File: src/components/state/loading.tsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { Component, h } from '@stencil/core';
+import { loading } from '../../utils/icon';
+
+@Component({
+  tag: 'xm-loading',
+  styleUrl: 'loading.scss',
+  shadow: true,
+})
+export class Loading {
```
[the remark this answers is not in the document]
Comment (christophblessing) [reply src/components/state/loading.tsx]: I changed the implementation accordingly. It reuses the `status` property to show / hide. The CSS is scoped to the player component.

File: src/components/state/loading.tsx [hunk 5]
```diff
@@ -1,29 +1,19 @@
-import { Component, h, Prop } from '@stencil/core';
-import { loading } from '../../utils/icon';
+import { FunctionalComponent, h } from '@stencil/core';
```
Comment (JuliaCasamitjana) [inline src/components/state/loading.tsx:1]: Another possible structure would be to move this file inside `/player` and name it `states.tsx.` From there we can export the `loading` state and add others in the future if needed.
It is how we do it for the `controls` component and all its functional components (`elements.tsx`).

Not a blocker, just an idea... also fine with your approach.

Comment (christophblessing) [reply src/components/state/loading.tsx:1]: I like that approach. We can take it even further: We also name it `element.tsx` and include video and controls content there. I would do the latter in a follow-up refactoring. This way, we can further reduce the size of the far to long `player.tsx` class.

**Review by JuliaCasamitjana — APPROVED** [review]

**Pull request merged**